src/docfx/lib/FileGlob.cs

```diff
@@ -0,0 +1,101 @@
+using System.IO;
+using System.Linq;
+
+namespace Microsoft.Docs
+{
+    public class FileGlob
+    {
+        public static IEnumerable<string> GetFiles(string cwd, IEnumerable<string> patterns, IEnumerable<string> excludePatterns, GlobMatcherOptions options = GlobMatcher.DefaultOptions)
+        {
+            // If there is no pattern, nothing will be included
+            if (patterns == null) return Enumerable.Empty<string>();
+            if (string.IsNullOrEmpty(cwd)) cwd = Directory.GetCurrentDirectory();
+
+            IEnumerable<GlobMatcher> globList = patterns.Select(s => new GlobMatcher(s, options));
+            IEnumerable<GlobMatcher> excludeGlobList = Enumerable.Empty<GlobMatcher>();
+            if (excludePatterns != null)
+                excludeGlobList = excludePatterns.Select(s => new GlobMatcher(s, options));
+            return GetFilesCore(cwd, globList, excludeGlobList);
+        }
+
+        private static IEnumerable<string> GetFilesCore(string cwd, IEnumerable<GlobMatcher> globs, IEnumerable<GlobMatcher> excludeGlobs)
+        {
+            if (!Directory.Exists(cwd)) yield break;
+            foreach (var file in GetFilesFromSubfolder(cwd, cwd, globs, excludeGlobs))
+            {
+                yield return NormalizeToFullPath(file);
+            }
+        }
+
+        private static IEnumerable<string> GetFilesFromSubfolder(string baseDirectory, string cwd, IEnumerable<GlobMatcher> globs, IEnumerable<GlobMatcher> excludeGlobs)
+        {
+            foreach (var file in Directory.GetFiles(baseDirectory, "*", SearchOption.TopDirectoryOnly))
+            {
+                var relativePath = GetRelativeFilePath(cwd, file);
+                if (IsFileMatch(relativePath, globs, excludeGlobs))
+                {
+                    yield return file;
+                }
+            }
+
+            foreach (var dir in Directory.GetDirectories(baseDirectory, "*", SearchOption.TopDirectoryOnly))
+            {
+                var relativePath = GetRelativeDirectoryPath(cwd, dir);
+                if (IsDirectoryMatch(relativePath, globs, excludeGlobs))
+                {
+                    foreach (var file in GetFilesFromSubfolder(dir, cwd, globs, excludeGlobs))
+                    {
+                        yield return file;
+                    }
+                }
+            }
+        }
+
+        private static string GetRelativeFilePath(string directory, string file)
+        {
+            var subpath = file.Substring(directory.Length);
+            // directory could be
+            // 1. root folder, e.g. E:\ or /
+            // 2. sub folder, e.g. a or a/ or a\
+            return subpath.TrimStart('\\', '/');
+        }
+
+        private static string GetRelativeDirectoryPath(string parentDirectory, string directory)
+        {
+            var relativeDirectory = GetRelativeFilePath(parentDirectory, directory);
+            return relativeDirectory.TrimEnd('\\', '/') + "/";
+        }
```

Comment:
why not just use Path.GetRelativePath()

Reply:
there is a bug in this function:
if I call `Path.GetRelativePath("test", "test/**.md")`, it returns `test/**.md` but not `**.md`

---
In reply to: [181018217](https://github.com/dotnet/docfx/pull/2627#discussion_r181018217) [](ancestors = 181018217)